feat: use v2 contracts

Nesopie opened the pull request.

Nesopie committed:
feat: use v2 contracts

kevinsslin commented on src/SelfPassportSBTV2.sol:
```diff
@@ -0,0 +1,219 @@
+    /// @param userData The user defined data passed through the verification process
+    function customVerificationHook(
+        ISelfVerificationRoot.GenericDiscloseOutputV2 memory genericDiscloseOutput,
+        bytes memory userData
```
if we are not using userData we can do it like this.

```solidity
bytes memory /* userData */
```

kevinsslin commented on src/SelfPassportSBTV2.sol:
```diff
@@ -0,0 +1,219 @@
+    /// @param userDefinedData The user defined data
+    /// @return The config ID
+    function getConfigId(
+        bytes32 destinationChainId,
```
```solidity
bytes32 /* destinationChainId */,
bytes32 /* userIdentifier */,
bytes memory /* userDefinedData */
```

kevinsslin commented on src/SelfPassportSBTV2.sol:
```diff
@@ -0,0 +1,219 @@
+        override
+    {
+        uint256 nullifier = genericDiscloseOutput.nullifier;
+        address receiver = address(uint160(genericDiscloseOutput.userIdentifier));
```
can we add receiver != address(0) validation here?

kevinsslin commented on src/SelfPassportSBTV2.sol:
```diff
@@ -0,0 +1,219 @@
+        _userToTokenId[tokenOwner] = 0;
+        delete _expiryTimestamps[tokenId];
+
+        // Need to find and clean up nullifier mappings
```
Are you going to implement this? @Nesopie

Nesopie replied:
Nah, the third "if" statement in customVerificationHook does nothing anyways if the token is burnt and the nullifier is not cleaned up. I'll just remove the comment

kevinsslin replied:
Case 1: Lost Passport ✅

  - Scenario: User loses physical/digital passport, gets new one
  - Result: Different nullifier generated from new passport
  - Recovery: Works perfectly
    - Admin burns old SBT → clears user mapping
    - New passport → new nullifier (not in _nullifierToTokenId)
    - Can mint fresh SBT to same wallet (Case 1: NEW nullifier + NO SBT)

  Case 2: Lost Private Key ❌

  - Scenario: User keeps same passport but loses wallet private key
  - Result: Same nullifier (same passport identity)
  - Recovery: Fails - fundamental design limitation
    - Admin burns SBT from old wallet → clears user mapping
    - User tries minting to new wallet with same nullifier
    - Triggers Case 3: USED nullifier + NO SBT → RegisteredNullifier() revert
    - Nullifier permanently tied to original wallet address

  Summary: Current burn function only supports passport loss recovery, not private key loss recovery.

@Nesopie I think we can add another mapping to cover the case 2. Wdyt?

Nesopie replied:
we can change the third if statement so that SBT gets transferred to the new address OR we can add another mapping that can clean up the nullifier (and leave the third if statement as it is)

kevinsslin replied:
I like the former one! Let's transfer to a new address. Can you update the README as well?

Nesopie committed:
fix: review changes

Nesopie committed:
lint

Nesopie committed:
feat: update post burn if condition

kevinsslin committed:
fix: implement proper Case 3 token recovery logic

- Case 3 now checks if token owner is address(0) for burned token recovery
- If token burned: recover to new address with same token ID
- If token active: revert to force admin burn first
- Updated burnSBT comments to clarify nullifier mapping preservation
- Enables secure wallet recovery while preventing nullifier hijacking

kevinsslin committed:
docs: add comprehensive recovery workflows and security analysis

- Added Recovery Workflows section with Lost Passport and Lost Wallet scenarios
- Updated Logic Matrix to reflect Case 3 RECOVER/REVERT behavior
- Added Case 3 breakdown table showing burned vs active token handling
- Documented Security Considerations including nullifier ambiguity attack
- Added mitigation strategies and acknowledged privacy vs security tradeoff
- Updated contract name references from V1 to V2

kevinsslin committed:
style: fix formatting issues from linter

- Run forge fmt to fix Solidity formatting
- Run prettier to fix README.md formatting
- All lint checks now pass
